python_modules/dagster/dagster/_core/definitions/sensor_definition.py

```diff
@@ -171,21 +151,56 @@ def __enter__(self) -> "SensorEvaluationContext":
         return self
 
     def __exit__(self, *exc) -> None:
-        self._resources_cm.__exit__(*exc)  # pylint: disable=no-member
+        if self._resources_cm is not None:
+            self._resources_cm.__exit__(*exc)  # pylint: disable=no-member
         self._exit_stack.close()
         self._logger = None
 
     def __del__(self) -> None:
-        if self._resources_contain_cm and not self._cm_scope_entered:
+        if (
+            self._resources_contain_cm
+            and not self._cm_scope_entered
+            and self._resources_cm is not None
+        ):
             self._resources_cm.__exit__(None, None, None)  # pylint: disable=no-member
 
     @property
     def resources(self) -> Resources:
+        if not self._resources:
+            """
+            This is similar to what we do in e.g. the op context - we set up a resource
+            building context manager, and immediately enter it. This is so that in cases
+            where a user is not using any context-manager based resources, they don't
+            need to enter this SensorEvaluationContext themselves.
+
+            For example:
+
+            my_sensor(build_sensor_context(resources={"my_resource": my_non_cm_resource})
+
+            will work ok, but for a CM resource we must do
+
+            with build_sensor_context(resources={"my_resource": my_cm_resource}) as context:
+                my_sensor(context)
+            """
+            from dagster._core.definitions.scoped_resources_builder import (
+                IContainsGenerator,
+            )
+            from dagster._core.execution.build_resources import build_resources
+
+            instance = self.instance if self._instance or self._instance_ref else None
+            self._resources_cm = build_resources(
+                resources=self._resource_defs or {}, instance=instance
+            )
+
+            self._resources = self._resources_cm.__enter__()
+
+            self._resources_contain_cm = isinstance(self._resources, IContainsGenerator)
+
```

Comment:
I think we can tidy stuff up here a bit - messing with `__del__` is cursed and its not clear to me that we need to.

The cases we have are:
* resource are context mangers
  - we have gone through `__enter__` and have an `ExitStack` we can append to we know will close on `__exit__`
  - we have not gone through `__enter__` and are about to throw `DagsterInvariantViolationError`, nothing to clean up since we are blowing up
* resources are not context managers
  - nothing to clean up 
  
So I think we can get rid of `__del__` and some of these other state vars and just use the exit stack and locals

Reply:
@alangenfeld would an argument for using __del__ be if we expect users to create their own context objects and can't neccesarily trust them to do so in contextmanagers?

Reply:
I think we can get rid of `__del__` since the two cases are:

1. User has no context manager resources. They are able to use the context directly as `my_sensor(build_sensor_context_manager(resources={"foo": bar}))`. There's no resource to clean up here.
2. User has context manager resources. Then, cleanup in del isn't needed since we'll call `__exit__`.